### PR TITLE
Check connection status before using it

### LINKED ISSUE #1754: master_add_node(<invalid_name>, 5432) crashes if there is a reference table
To reproduce;
```SQL
create table test(i int);
select create_reference_table('test');
SELECT * from master_add_node('node-name', 5432);
```

back trace;
```
#0  0x00007fffb31eed42 in __pthread_kill ()
#1  0x00007fffb32dc457 in pthread_kill ()
#2  0x00007fffb3154420 in abort ()
#3  0x0000000104daa0c0 in ExceptionalCondition (conditionName=0x105378cb7 "!(PQisnonblocking(pgConn))", errorType=0x1053771c1 "FailedAssertion", fileName=0x105378cd2 "connection/remote_commands.c", lineNumber=465) at assert.c:54
#4  0x00000001052e5ef0 in SendRemoteCommand (connection=0x7f971f81be68, command=0x7f971f8536f8 "BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;SELECT assign_distributed_transaction_id(0, 1, '2017-11-01 17:26:13.240799+03');") at connection/remote_commands.c:465
#5  0x000000010533a682 in StartRemoteTransactionBegin (connection=0x7f971f81be68) at transaction/remote_transaction.c:102
#6  0x000000010533a915 in RemoteTransactionBegin (connection=0x7f971f81be68) at transaction/remote_transaction.c:143
#7  0x000000010533e7c7 in SendCommandListToWorkerInSingleTransaction (nodeName=0x7f971f82b440 "node-name", nodePort=5432, nodeUser=0x7f971f853640 "mehmetfurkansahin", commandList=0x7f971f852e88) at transaction/worker_transaction.c:226
#8  0x000000010535679a in ReplicateShardToNode (shardInterval=0x7f971f82cbc0, nodeName=0x7f971f82b440 "node-name", nodePort=5432) at utils/reference_table_utils.c:298
#9  0x0000000105356500 in ReplicateAllReferenceTablesToNode (nodeName=0x7f971f82b440 "node-name", nodePort=5432) at utils/reference_table_utils.c:156
#10 0x0000000105351861 in ActivateNode (nodeName=0x7f971f82b440 "node-name", nodePort=5432) at utils/node_metadata.c:449
#11 0x00000001053514ac in master_add_node (fcinfo=0x7fff5b4c68d0) at utils/node_metadata.c:137
#12 0x00000001049f7732 in ExecMakeTableFunctionResult (setexpr=0x7f971f83fc78, econtext=0x7f971f83f970, argContext=0x7f9721809168, expectedDesc=0x7f971f826940, randomAccess=0 '\0') at execSRF.c:231
#13 0x0000000104a0b78e in FunctionNext (node=0x7f971f83f858) at nodeFunctionscan.c:95
#14 0x00000001049f689b in ExecScanFetch (node=0x7f971f83f858, accessMtd=0x104a0b6e0 <FunctionNext>, recheckMtd=0x104a0bb00 <FunctionRecheck>) at execScan.c:97
#15 0x00000001049f659c in ExecScan (node=0x7f971f83f858, accessMtd=0x104a0b6e0 <FunctionNext>, recheckMtd=0x104a0bb00 <FunctionRecheck>) at execScan.c:147
#16 0x0000000104a0b322 in ExecFunctionScan (pstate=0x7f971f83f858) at nodeFunctionscan.c:270
#17 0x00000001049f4a62 in ExecProcNodeFirst (node=0x7f971f83f858) at execProcnode.c:430
#18 0x00000001049f17a2 in ExecProcNode (node=0x7f971f83f858) at executor.h:250
#19 0x00000001049ed536 in ExecutePlan (estate=0x7f971f83f640, planstate=0x7f971f83f858, use_parallel_mode=0 '\0', operation=CMD_SELECT, sendTuples=1 '\001', numberTuples=0, direction=ForwardScanDirection, dest=0x7f971f8461d0, execute_once=1 '\001') at execMain.c:1721
#20 0x00000001049ed40b in standard_ExecutorRun (queryDesc=0x7f971f846840, direction=ForwardScanDirection, count=0, execute_once=1 '\001') at execMain.c:363
#21 0x00000001049ed1c7 in ExecutorRun (queryDesc=0x7f971f846840, direction=ForwardScanDirection, count=0, execute_once=1 '\001') at execMain.c:306
#22 0x0000000104bf6b72 in PortalRunSelect (portal=0x7f971f05bc40, forward=1 '\001', count=0, dest=0x7f971f8461d0) at pquery.c:932
#23 0x0000000104bf653a in PortalRun (portal=0x7f971f05bc40, count=9223372036854775807, isTopLevel=1 '\001', run_once=1 '\001', dest=0x7f971f8461d0, altdest=0x7f971f8461d0, completionTag=0x7fff5b4c7230 "") at pquery.c:773
#24 0x0000000104bf1ebc in exec_simple_query (query_string=0x7f971f037640 "SELECT * from master_add_node('node-name', 5432);") at postgres.c:1099
#25 0x0000000104bf10c7 in PostgresMain (argc=1, argv=0x7f9720009bf8, dbname=0x7f9720009a58 "postgres", username=0x7f971f002fc8 "mehmetfurkansahin") at postgres.c:4090
#26 0x0000000104b3740e in BackendRun (port=0x7f971ed11a30) at postmaster.c:4357
#27 0x0000000104b36793 in BackendStartup (port=0x7f971ed11a30) at postmaster.c:4029
#28 0x0000000104b35775 in ServerLoop () at postmaster.c:1753
#29 0x0000000104b32f61 in PostmasterMain (argc=5, argv=0x7f971ec02c60) at postmaster.c:1361
#30 0x0000000104a56eb9 in main (argc=5, argv=0x7f971ec02c60) at main.c:228
```

### PR DESCRIPTION
**Pull request by metdos**
Fixes #1754

### CODE REVIEW

File: src/test/regress/expected/multi_replicate_reference_table.out [hunk 1]
```diff
@@ -713,6 +713,12 @@ SELECT 1 FROM master_add_node('localhost', :worker_2_port);
         1
 (1 row)
 
+-- test adding an invalid node while we have reference tables to replicate
+-- set client message level to ERROR to suppress OS depended host name resolution warnings
```
Comment (marcocitus) [inline src/test/regress/expected/multi_replicate_reference_table.out]: Not worth fixing, but it's "OS-dependent"

File: src/test/regress/expected/multi_router_planner.out [hunk 2]
```diff
@@ -2151,7 +2151,7 @@ GRANT INSERT ON ALL TABLES IN SCHEMA public TO router_user;
 BEGIN;
 INSERT INTO failure_test VALUES (1, 1);
 WARNING:  connection error: localhost:57638
-DETAIL:  no connection to the server
+DETAIL:  FATAL:  role "router_user" does not exist
```
Comment (marcocitus) [inline src/test/regress/expected/multi_router_planner.out]: Why did this message change?

Comment (marcocitus) [reply src/test/regress/expected/multi_router_planner.out]: (I like that it did, just wondering what changed about the code path)

Comment (metdos) [reply src/test/regress/expected/multi_router_planner.out]: It looks like, we used to try sending the query even if the connection was not okay. Then, this was overwriting the error message in the `conn->errorMessage.data;` and we were getting `no connection to the server` error.

Here is the call stack
```
SendRemoteCommand() at remote_commands.c:462 0x10457ade0	
StartRemoteTransactionBegin() at remote_transaction.c:102 0x1045cba02	
RemoteTransactionsBeginIfNecessary() at remote_transaction.c:601 0x1045cc9b3	
GetModifyConnections() at multi_router_executor.c:926 0x104581ddd	
ExecuteSingleModifyTask() at multi_router_executor.c:731 0x1045806d6	
RouterSequentialModifyExecScan() at multi_router_executor.c:484 0x10458053d	
ExecCustomScan() at nodeCustom.c:118 0x10384b3e8	
ExecProcNodeFirst() at execProcnode.c:430 0x103835842	
ExecProcNode() at executor.h:250 0x1038325b2	
ExecutePlan() at execMain.c:1,721 0x10382e346	
standard_ExecutorRun() at execMain.c:363 0x10382e21b	
ExecutorRun() at execMain.c:306 0x10382dfd7	
ProcessQuery() at pquery.c:161 0x103a3ad1e	
PortalRunMulti() at pquery.c:1,286 0x103a39d88	
PortalRun() at pquery.c:799 0x103a39464	
exec_simple_query() at postgres.c:1,099 0x103a34cec	
PostgresMain() at postgres.c:4,090 0x103a33ef7	
BackendRun() at postmaster.c:4,357 0x10397ae5e	
BackendStartup() at postmaster.c:4,029 0x10397a1d3	
ServerLoop() at postmaster.c:1,753 0x1039791c5	
<...more frames...>	
```

**Comment by metdos** [comment]
@marcocitus anything blocking this PR to be merged?

**Pull request merged**